File: Algorithms/Easy/ReplaceElementsWithGreatestElementOnRightSide.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def replaceElements(self, arr: list[int]) -> list[int]:
        curr_max = -1

        for i in range(len(arr) - 1, -1, -1):
            new_max = max(curr_max, arr[i])
            arr[i] = curr_max
            curr_max = new_max

        return arr

File: Algorithms/Easy/ReplaceElementsWithGreatestElementOnRightSide.go
package easy

// Time complexity: O(n)
// Space complexity: O(1)
func replaceElements(arr []int) []int {
	maxVal := -1
	for i := len(arr) - 1; i > -1; i-- {
		temp := arr[i]
		arr[i] = maxVal
		maxVal = max(temp, maxVal)
	}

	return arr
}
